### ISSUE TITLE
`instanceof` doesn't work in middleware v12.2

### ISSUE THREAD

**Issue by meikidd** [vercel/next.js#38184]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64
    Binaries:
      Node: 17.8.0
      npm: 8.5.5
      Yarn: 1.22.18
      pnpm: N/A
    Relevant packages:
      next: 12.2.0
      eslint-config-next: 12.2.0
      react: 18.2.0
      react-dom: 18.2.0

### Describe the Bug

The `instanceof` operator doesn't work in the new released middleware.

```typescript
export async function middleware(request: NextRequest) {
  const response = NextResponse.next();

  const obj = { hello: "world" };
  console.log(obj instanceof Object); // <-- this prints "false" in new middleware

  return response;
}
```

### Expected Behavior

Expect above example to print "true".

### Link to reproduction

https://github.com/meikidd/nextjs-middleware-instanceof

### To Reproduce

1. npm run dev
2. Open "http://localhost:3000"
3. Check the logs in Terminal. We expect "true", but actually get "false".

**Comment by feugy** [comment]
Thanks you @meikidd.
Indeed, it is an issue we have **in dev only**. 

Calling `instanceof` on literals will fail because we're running the code in a sandbox that has different primitives as Node.js:

```js
// will return false in the middleware
[] instanceof Array
{} instanceof Object
/^hello$/ instanceof RegExp
```

We're working on a fix, but it's a tricky problem, due to the nature of [Node.js VM module](https://github.com/nodejs/help/issues/3907).

**Comment by SukkaW** [comment]
@feugy FYI, there used to be a dirty hack for this:

```js
Array = [].constructor;
```

**Comment by feugy** [comment]
That is... fishy indeed 😅 
But it'll be greatly useful. Let me run more testing on our end. Thanks @SukkaW !!

**Comment by SukkaW** [comment]
@feugy If the issue **only occurs in dev mode**, IMHO the hacky workaround should work.

Also, per [Node.js' `vm` API Documentation](https://nodejs.org/dist/latest-v18.x/docs/api/vm.html):

> The `node:vm` module is not a security mechanism. Do not use it to run untrusted code.

Although it is only used in **local developing environment**, it is still technically not safe, and also causes behavior mismatches between development and production (just like this issue).

But there is an escape hatch. There are ongoing ECMAScript proposals [proposal-compartments](https://github.com/tc39/proposal-compartments) (currently at stage 1) and [proposal-shadowrealm](https://github.com/tc39/proposal-shadowrealm) (currently at stage 3 already!) to help developer creating a safe, secure sandbox for executing code!

**Comment by feugy** [comment]
Thanks for the pointers on the incoming ECMA proposals. As you said it, edge-runtime is currently for dev only, and is not meant to provide strong isolation. We'll surely revise is in the future, but it is out of scope for now.

We did some investigation with @javivelasco about the hack, and unfortunately it's not gonna work.

<details>
<summary>
TLDR; it's fixing <code>instanceof</code>, but breaking the polyfills
</summary>
<div>
<br/>

The EdgeVM (node.js's VM module under the hood), has its own _realm_ with its own `Array` constructor.
Since it's just a plain JavaScript environment, we must load some polyfills into it, to provide utilities such as `TextEncoder`, `FormData` and `Fetch` (there are dozens) of [them](https://github.com/vercel/edge-runtime/blob/main/packages/primitives/src/index.js).

In the current implementation, we load the polyfills in node.js _realm_, and [injects them into the EdgeVM as globals](https://github.com/vercel/edge-runtime/blob/main/packages/vm/src/edge-vm.ts#L35).
But there's an drawback: the `Array` constructor used in the polyfill is the comming from node.js realm, while the EdgeVM has its own one.
To fix this inconsistency, we also inject node.js' `Array` into the EdgeVM context.

However, the JS engine **does not use global `Array` constructor while building an array literal**. And there's no way to change this, because it's how the ECMAscript spec defines it. So we have an inconsistency leading to this bug.

The suggested hack: `Array = [].constructor` reverts it back: the `Array` constructor inside EdgeVM realm becomes the same as the array literal, and different from the one used in the polyfills.
</div>
</details>

---

The only way to definitely fix this is to stop injecting anything from node.js, and load all the polyfills within the EdgeVM.
It is nowhere simple, because EdgeVM (node.js's VM module) is just a barebone JavaScript engine. It has no `Buffer`, no `require()`... And we need to provide or implement them.

**Comment by Kikobeats** [comment]
Hello, this has been addressed at Next.js v12.2.3:

```js
console.log(request instanceof Object) // true
console.log(response instanceof Object) // true
console.log({} instanceof Object) // true
```

You can see more examples [Edge Runtime test suite](https://github.com/vercel/edge-runtime/blob/main/packages/vm/tests/unit/edge-vm.test.ts#L171=) 🙂
